### PR TITLE
Ascalva/roles

### PR DESCRIPTION
**Pull request by ascalva**
Create views and roles for users and admins

### CODE REVIEW

File: src/main/java/rpgcharacters/RpgCharacters.java [hunk 1]
```diff
@@ -78,7 +76,40 @@ public static void main(String[] args) {
...
+
+            String userView = "CREATE VIEW userView AS "
+                            + "SELECT * "
+                            + "FROM party, character;";
+            s.executeQuery(userView);
+
+            //Change later
+            String adminView = "CREATE VIEW adminView AS "
+                             + "SELECT * "
+                             + "FROM *;";
```
Comment (ascalva) [inline src/main/java/rpgcharacters/RpgCharacters.java]: Forgot to write this part, update soon

File: src/main/java/rpgcharacters/RpgCharacters.java [hunk 2]
```diff
@@ -78,7 +76,40 @@ public static void main(String[] args) {
             jc.usage();
             return;
         }
+        //Create Roles
+        try {
+            Statement s = conn.createStatement();
+
+            String userView = "CREATE VIEW userView AS "
+                            + "SELECT * "
+                            + "FROM party, character;";
```
Comment (ekliot) [inline src/main/java/rpgcharacters/RpgCharacters.java]: will this mean a user cannot see quests or items, even if it is connected to their party or their character?

File: src/main/java/rpgcharacters/RpgCharacters.java [hunk 3]
```diff
@@ -1,15 +1,13 @@
...
-import java.sql.SQLException;
+import java.sql.*;
 import java.util.Hashtable;
 
 import com.beust.jcommander.JCommander;
 import com.beust.jcommander.Parameter;
 
 import rpgcharacters.commands.*;
-import rpgcharacters.userflow.*;
+import rpgcharacters.userFlow.*;
```
Comment (ekliot) [inline src/main/java/rpgcharacters/RpgCharacters.java]: undo this, the package is `rpgcharacters.userflow`

File: src/main/java/rpgcharacters/commands/CommandCreateUser.java [hunk 4]
```diff
@@ -30,6 +30,19 @@ public void run(Connection conn) {
...
+            query = "GRANT dbUser "
+                  + "TO (SELECT username "
+                  +     "FROM user "
+                  +     "WHERE username = '" + this.username + "');";
+            stmt.executeQuery(query);
+
+            query = "GRANT dbAdmin "
+                  + "TO (SELECT admin "
+                  +     "FROM user "
+                  +     "WHERE admin = TRUE);";
```
Comment (ekliot) [inline src/main/java/rpgcharacters/commands/CommandCreateUser.java]: this query should not be run every time on user creation, either... admin role is granted to the first user in the db, and then only to those users that user grants admin privileges to. replace this with a `SELECT COUNT(*) AS userCount ...`, and then if `userCount == 1` AFTER the user has been created, grant dbAdmin to that user

File: src/main/java/rpgcharacters/RpgCharacters.java (hunk 2 above)
Comment (returningsam) [inline src/main/java/rpgcharacters/RpgCharacters.java]: Also I think the creation of the roles should happen in the create tables command

File: src/main/java/rpgcharacters/RpgCharacters.java (hunk 1 above)
Comment (returningsam) [inline src/main/java/rpgcharacters/RpgCharacters.java]: these as well

File: src/main/java/rpgcharacters/commands/CommandCreateUser.java [hunk 5]
```diff
@@ -30,6 +30,19 @@ public void run(Connection conn) {
...
+                  +     "FROM user "
+                  +     "WHERE username = '" + this.username + "');";
+            stmt.executeQuery(query);
+
+            query = "GRANT dbAdmin "
+                  + "TO (SELECT COUNT( * ) AS userCount "
+                  +     "FROM user "
+                  +     "WHERE userCount = 1);";
+            stmt.executeQuery(query);
+
```
Comment (mwitte98) [inline src/main/java/rpgcharacters/commands/CommandCreateUser.java]: I deleted this file in my PR so this code for granting roles will need to be moved to CreateUserMenu. I merged master into this branch so the SQL is there.

File: src/main/java/rpgcharacters/commands/CommandCreateUser.java [hunk 6]
```diff
@@ -30,6 +30,19 @@ public void run(Connection conn) {
...
+            query = "GRANT dbUser "
+                  + "TO (SELECT username "
+                  +     "FROM user "
+                  +     "WHERE username = '" + this.username + "');";
+            stmt.executeQuery(query);
+
+            query = "GRANT dbAdmin "
+                  + "TO (SELECT COUNT( * ) AS userCount "
+                  +     "FROM user "
+                  +     "WHERE userCount = 1);";
```
Comment (mwitte98) [inline src/main/java/rpgcharacters/commands/CommandCreateUser.java]: I don't think this is valid SQL. We'll probably have to do `SELECT COUNT(*) as userCount FROM user` first. Then do something like
```
if (results.getInt("userCount") == 1) {
    // grant admin to user
} else {
    // grant db user to user
}
```

**Pull request merged**